src/gfx.cpp

```diff
@@ -1445,20 +1438,75 @@ void RedrawScreenRect(int left, int top, int right, int bottom)
 	VideoDriver::GetInstance()->MakeDirty(left, top, right - left, bottom - top);
 }
 
+static std::vector<Rect> _dirty_viewport_occlusions;
+static const Viewport *_dirty_viewport;
+static NWidgetDisplay _dirty_viewport_disp_flags;
+
+static void DrawDirtyViewport(uint occlusion, int left, int top, int right, int bottom)
+{
+	for(; occlusion < _dirty_viewport_occlusions.size(); occlusion++) {
+		const Rect &occ = _dirty_viewport_occlusions[occlusion];
+		if (right <= occ.left || bottom <= occ.top || left >= occ.right || top >= occ.bottom) {
+			/* No intersection. */
+			continue;
+		}
+
+		int middle;
+		if (left < (middle = occ.left)) {
+			/* Occlusion intersects with dirty rect; split at left edge. */
+			DrawDirtyViewport(occlusion + 1, left, top, middle, bottom);
+			DrawDirtyViewport(occlusion, middle, top, right, bottom);
+			return;
+		}
+
+		if (right > (middle = occ.right)) {
+			/* Occlusion intersects with dirty rect; split at right edge. */
+			DrawDirtyViewport(occlusion, left, top, middle, bottom);
+			DrawDirtyViewport(occlusion + 1, middle, top, right, bottom);
+			return;
+		}
+
+		if (top < (middle = occ.top)) {
+			/* Occlusion intersects with dirty rect; split at top edge. */
+			DrawDirtyViewport(occlusion + 1, left, top, right, middle);
+			DrawDirtyViewport(occlusion, left, middle, right, bottom);
+			return;
+		}
+
+		if (bottom > (middle = occ.bottom)) {
+			/* Occlusion intersects with dirty rect; split at bottom edge. */
+			DrawDirtyViewport(occlusion, left, top, right, middle);
+			DrawDirtyViewport(occlusion + 1, left, middle, right, bottom);
+			return;
+		}
+
+		/* Dirty rect is completely occluded. Nothing more to do. */
+		return;
+	}
+
+	if (_game_mode == GM_MENU) {
+		RedrawScreenRect(left, top, right, bottom);
+	} else {
+		extern void ViewportDrawChk(const Viewport *vp, int left, int top, int right, int bottom);
+		ViewportDrawChk(_dirty_viewport, left, top, right, bottom);
```

Comment:
Given the apparent purpose of `ViewportDrawChk`; i.e. to avoid overflowing the sprite cache; is there any reason why the menu screen needs special treatment here?
The only one I can think of is that the menu screen is not zoomable; but with 8K monitors being a thing, perhaps the assumption being made here will soon become invalid.

Reply:
It's because of the OpenTTD logo: https://github.com/OpenTTD/OpenTTD/blob/master/src/main_gui.cpp#L279-L293

Reply:
That seems like a good enough reason. How big does the monitor have to be before the sprite cache will overflow?